Patch from ticket.

adamsilverstein opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket:  https://core.trac.wordpress.org/ticket/39941

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

johnbillion commented on src/wp-includes/functions.php:
```diff
@@ -7781,3 +7781,123 @@ function is_php_version_compatible( $required ) {
+				$attributes_string .= $html5_script_support ? sprintf( ' %1$s="%2$s"', $attribute_name, esc_attr( $attribute_name ) ) : ' ' . $attribute_name;
+			}
+		} else {
+			$attributes_string .= sprintf( ' %1$s="%2$s"', $attribute_name, esc_attr( $attribute_value ) );
```
Given this function has `sanitize` in its name I would expect the attribute names to be sanitised as well as the values, but the names are not. I could see a situation where a plugin passes untrusted user input in to the `$attributes` parameter and the attribute name ends up being a vulnerability.

adamsilverstein replied:
I think this may have been in place before; will add `esc_attr` here.

johnbillion commented on src/wp-includes/functions.php:
```diff
@@ -7781,3 +7781,123 @@ function is_php_version_compatible( $required ) {
+	 *                          Only the attribute name is added to the `<script>` tag for
+	 *                          entries with a boolean value, and that are true.
+	 */
+	$attributes = apply_filters( 'wp_script_attributes', $attributes );
```
Should the `$javascript` parameter be passed to this filter too? It's the only other context available.

johnbillion replied:
Whoops this was meant for the filter in the `wp_get_inline_script_tag()` function. If it is added, there will be two definitions of the filter. Worth them being separate filters instead of the same?

adamsilverstein replied:
Probably a good idea to separate them for maximum flexibility, remaining the second `wp_inline_script_attributes` and adding $javascript for context.

johnbillion commented:
In general this seems fine, except the comment about the attribute names. That said, it would be good to see these changes in the context of where the functions will be used, eg. in the script loader tags, CSP-protected inline JS, etc.

adamsilverstein commented on src/wp-includes/functions.php:
```diff
@@ -7781,3 +7781,123 @@ function is_php_version_compatible( $required ) {
+/**
+ * Sanitizes an attributes array into an attributes string to be placed inside a `<script>` tag.
+ *
+ * Automatically injects type attribute if needed.
+ * Used by {@see wp_get_script_tag()} and {@see wp_get_inline_script_tag()}.
```
removed

adamsilverstein commented on tests/phpunit/tests/functions/wpInlineScriptTag.php:
```diff
@@ -0,0 +1,119 @@
+});
+JS;
+
+	function get_inline_script_tag_type_set() {
```
Added.

adamsilverstein commented:
> In general this seems fine, except the comment about the attribute names. That said, it would be good to see these changes in the context of where the functions will be used, eg. in the script loader tags, CSP-protected inline JS, etc.

@johnbillion I addressed your feedback here. In terms of how these will be used, take a look at  https://core.trac.wordpress.org/ticket/39941 / https://github.com/WordPress/wordpress-develop/pull/498) and  https://core.trac.wordpress.org/ticket/51407 / https://github.com/WordPress/wordpress-develop/pull/551 which are the follow up issues created to actually use this new API.

hellofromtonya commented:
Closed with changeset https://core.trac.wordpress.org/changeset/50167
